File: srcp/day-6.py
data = """
....#.....
.........#
..........
..#.......
.......#..
..........
.#..^.....
........#.
#.........
......#...
"""
import time

start = time.time()

data = data.strip().splitlines()
grid: list[list[str]] = list(map(list, data))

turnOrder: list[str] = ['^', '>', 'v', '<']
positionInitGuard: list[int] = [0, 0]
directionInitGuard: str = '^'

for row in range(len(grid)):
    for col in range(len(grid[0])):
        cell = grid[row][col]
        if cell in turnOrder:
            positionInitGuard[0] = row
            positionInitGuard[1] = col
            directionInitGuard = cell
            break


def simulate_guard_patrol(grid: list[list[str]], rows: int, cols: int,
                          ini_position: list[int], init_dir: str,
                          detect_loop: bool = False) -> tuple[bool, set[str]]:
    directions = {
        '^': (-1, 0),
        '>': (0, 1),
        'v': (1, 0),
        '<': (0, -1),
    }

    turn_order = ['^', '>', 'v', '<']
    guard_pos = list(ini_position)
    guard_dir = init_dir

    visited = set()

    while True:
        r, c = guard_pos
        state = f"{r},{c},{guard_dir}" if detect_loop else f"{r},{c}"

        # Vérification de boucle
        if state in visited and detect_loop:
            return True, visited

        visited.add(state)

        dr, dc = directions[guard_dir]
        fr, fc = r + dr, c + dc

        if fr < 0 or fr >= rows or fc < 0 or fc >= cols:
            return False, visited

        front_cell = grid[fr][fc]

        if front_cell == "#" or front_cell == "O":
            current_dir_idx = turn_order.index(guard_dir)
            guard_dir = turn_order[(current_dir_idx + 1) % len(turn_order)]
        else:
            guard_pos = [fr, fc]


q_row: int = len(grid)
q_col: int = len(grid[0])
time_int = time.time() - start
totalPart1: [bool, set[str]] = simulate_guard_patrol(grid, q_row, q_col, positionInitGuard,directionInitGuard)
print(f"Part 1: {len(totalPart1[1])} en {time_int:.2f}s")

totalPart2: int = 0

for cell in totalPart1[1]:
    r, c = cell.split(",")
    if grid[int(r)][int(c)] == ".":
        grid[int(r)][int(c)] = "O"
        (is_loop, visited) = simulate_guard_patrol(grid, q_row, q_col, positionInitGuard, directionInitGuard, True)
        if is_loop:
            totalPart2 += 1
        grid[int(r)][int(c)] = "."


end = time.time()
print(f"Part 2: {totalPart2} en {(time_int + (end - start)):.2f}s")

File: srcr/src/days_2024/day6.rs
use std::collections::HashSet;
use std::time::Instant;
use rayon::prelude::*;


type Direction = char;
const TURN_ORDER: [Direction; 4] = ['^', '>', 'v', '<'];

fn get_init_pos(grid: &Vec<Vec<char>>) -> ([usize; 2], Direction) {
    let rows = grid.len();
    let cols = grid[0].len();

    for r in 0..rows {
        for c in 0..cols {
            if TURN_ORDER.contains(&grid[r][c]) {
                return ([r, c], grid[r][c]);
            }
        }
    }
    panic!("Guard position or direction not found on the map");
}

fn simulate_guard_patrol(
    grid: &Vec<Vec<char>>,
    init_pos: [usize; 2],
    init_dir: Direction,
    detect_loop: bool,
) -> (bool, HashSet<String>) {
    let directions = [
        ('^', (-1, 0)),
        ('>', (0, 1)),
        ('v', (1, 0)),
        ('<', (0, -1)),
    ]
        .iter()
        .cloned()
        .collect::<std::collections::HashMap<_, _>>();

    let mut guard_pos = init_pos;
    let mut guard_dir = init_dir;
    let mut visited = HashSet::new();

    loop {
        let (r, c) = (guard_pos[0], guard_pos[1]);
        let state = if detect_loop {
            format!("{},{},{}", r, c, guard_dir)
        } else {
            format!("{},{}", r, c)
        };

        if visited.contains(&state) && detect_loop {
            return (true, visited);
        }

        visited.insert(state);

        let (dr, dc) = directions[&guard_dir];
        let (fr, fc) = (r as isize + dr, c as isize + dc);

        if fr < 0
            || fr as usize >= grid.len()
            || fc < 0
            || fc as usize >= grid[0].len()
        {
            return (false, visited);
        }

        let front_cell = grid[fr as usize][fc as usize];

        if front_cell == '#' || front_cell == 'O' {
            let current_dir_idx = TURN_ORDER.iter().position(|&d| d == guard_dir).unwrap();
            guard_dir = TURN_ORDER[(current_dir_idx + 1) % 4];
        } else {
            guard_pos = [fr as usize, fc as usize];
        }
    }
}

pub fn solve_day_6() {
    let data_day_6 = "
grid
    ";

    let start_time: Instant = Instant::now();

    let grid: Vec<Vec<char>> = data_day_6.trim().to_string()
        .lines()
        .map(|line| line.trim().chars().collect())
        .collect();

    let (guard_init_pos, guard_init_dir) = get_init_pos(&grid);

    let ( _ , visited) =
        simulate_guard_patrol(&grid, guard_init_pos, guard_init_dir, false);

    let time_end_part1 = Instant::now();
    println!(
        "Partie 1 : {} , en : {:.2}s",
        visited.len(),
        (time_end_part1 - start_time).as_secs_f64()
    );

    let total_part3: usize = visited.par_iter()
        .map(|cell| {
            let mut split = cell.split(',');
            let r: usize = split.next().unwrap().parse().unwrap();
            let c: usize = split.next().unwrap().parse().unwrap();

            if grid[r][c] == '.' {
                let mut tmp_grid = grid.clone();
                tmp_grid[r][c] = 'O';
                let (looping, _) = simulate_guard_patrol(&grid, guard_init_pos, guard_init_dir, true);
                if looping { 1 } else { 0 }
            } else {
                0
            }
        })
        .sum();

    let time_end_part2: Instant = Instant::now();
    println!(
        "Partie 2 : {} en : {:.2}s",
        total_part3,
        (time_end_part2 - time_end_part1).as_secs_f64()
    );
}